templates/templates_test.go

```diff
@@ -89,3 +89,56 @@ func TestParseTruncateFunction(t *testing.T) {
 		})
 	}
 }
+
+func TestHeaderFunctions(t *testing.T) {
+	const source = "hello world"
+
+	tests := []struct {
+		doc      string
+		template string
+		expected string
+	}{
+		{
+			doc:      "json",
+			template: `{{ json .}}`,
+		},
+		{
+			doc:      "split",
+			template: `{{ split . ","}}`,
+		},
+		{
+			doc:      "join",
+			template: `{{ join . ","}}`,
+		},
+		{
+			doc:      "title",
+			template: `{{ title .}}`,
+		},
+		{
+			doc:      "lower",
+			template: `{{ lower .}}`,
+		},
+		{
+			doc:      "upper",
+			template: `{{ upper .}}`,
+		},
+		{
+			doc:      "truncate",
+			template: `{{ truncate . 2}}`,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.doc, func(t *testing.T) {
+			tmpl, err := New("").Funcs(HeaderFunctions).Parse(tc.template)
+			assert.NilError(t, err)
+
+			var b bytes.Buffer
+			assert.NilError(t, tmpl.Execute(&b, source))
+
+			// All header-functions are currently stubs, and don't modify the input.
+			expected := source
+			assert.Equal(t, expected, b.String())
```

Comment:
how can this test pass if expected isn't defined inside our test table? We are just comparing the source text to whatever the template produced, which can't be the same in all instances, right?

Reply:
Yes, they _do_ produce the same output as the input; it's a bit of a hack.

When using the `table` format, the user specifies the template for rows; the same template is used for printing the columns (so that, say, swapping the order of columns in the template will also swap the table headers).

However, part of the template could be to (e.g.) truncate the output, uppercase/lowercase it, etc, and we don't want the table headers to be changed by those, so the `HeaderFunctions` override some of the existing functions (by default `basicFunctions`) with stub functions so that they don't take effect on the table headers.


```bash
docker image ls --format 'table {{upper .Repository}}:{{ .Tag}}\t{{lower .ID}}'
REPOSITORY:TAG               IMAGE ID
DOCKER-CLI-DEV:latest        0df2b0f75f92
DOCKER-DEV:latest            339109b3c24f
UBUNTU:24.04                 ab64a8382e93
NGINX:latest                 d2eb56950b84
NGINX:alpine                 2140dad235c1
```

Reply:
I initially had it in the test-table, but moved the `expected` variable here so that I could put a single comment to outline it's expected to always be the same for all test-cases (at least currently). But happy to add back `expected` in the test-table itself.

Reply:
ah okay. It seemed like the `expected` value inside the table was forgotten since it was declared https://github.com/docker/cli/pull/5541/files#diff-c6842e8d411be4977abca67d350e55f1cb1a144154846487c956c4ef172e6a17R99

I would be fine to keep it like this and remove the defined `expected` inside the test table.

Reply:
oh!! Yes, that's wrong; looks like I forgot to remove it there, thought I did; let me fix that part 🙈

Reply:
Done; removed it from the table 😅